pdr_backend/trader/trader_agent.py

```diff
@@ -81,11 +81,19 @@ def take_step(self):
         self.prev_block_number = block_number
         self.prev_block_timestamp = block["timestamp"]
 
+        s_till_epoch_ends = []
+
         # do work at new block
         for addr in self.feeds:
-            self._process_block_at_feed(addr, block["timestamp"])
+            s_till_epoch_ends.append(
+                self._process_block_at_feed(addr, block["timestamp"])
+            )
+
+        sleep_time = min(s_till_epoch_ends) - 1
+        print(f"-- Soonest epoch is in {sleep_time} seconds, waiting... --")
+        time.sleep(sleep_time)
 
-    def _process_block_at_feed(self, addr: str, timestamp: int, tries: int = 0):
+    def _process_block_at_feed(self, addr: str, timestamp: int, tries: int = 0) -> int:
```

Comment:
If this method is returning values, there should be a docstring of what it's returning.